src/librustdoc/html/render/span_map.rs

```diff
@@ -0,0 +1,209 @@
+use crate::clean;
+use crate::html::sources;
+
+use rustc_data_structures::fx::FxHashMap;
+use rustc_hir::def::{DefKind, Res};
+use rustc_hir::def_id::DefId;
+use rustc_hir::intravisit::{self, NestedVisitorMap, Visitor};
+use rustc_hir::{ExprKind, GenericParam, GenericParamKind, HirId, Mod, Node};
+use rustc_middle::ty::TyCtxt;
+use rustc_span::Span;
+
+use std::path::{Path, PathBuf};
+
+/// This enum allows us to store two different kinds of information:
+///
+/// In case the `span` definition comes from the same crate, we can simply get the `span` and use
+/// it as is.
+///
+/// Otherwise, we store the definition `DefId` and will generate a link to the documentation page
+/// instead of the source code directly.
+#[derive(Debug)]
+crate enum LinkFromSrc {
+    Local(clean::Span),
+    External(DefId),
+}
+
+/// This struct is used only as index in the `span_map`, not as [`Span`]! `Span`s contain
+/// some extra information (the syntax context) we don't need. **Do not convert this type back to
+/// `Span`!!!**
+#[derive(Clone, Copy, Hash, PartialEq, Eq, Debug)]
+crate struct LightSpan {
+    crate lo: u32,
+    crate hi: u32,
+}
+
+impl LightSpan {
+    /// Before explaining what this method does, some global explanations on rust's `Span`:
+    ///
+    /// Each source code file is stored in the source map in the compiler and has a
+    /// `lo` and a `hi` (lowest and highest bytes in this source map which can be seen as one huge
+    /// string to simplify things). So in this case, this represents the starting byte of the
+    /// current file. It'll be used later on to retrieve the "definition span" from the
+    /// `span_correspondance_map` (which is inside `context`).
+    ///
+    /// This when we transform the "span" we have from reading the input into a "span" which can be
+    /// used as index to the `span_correspondance_map` to get the definition of this item.
+    ///
+    /// So in here, `file_span_lo` is representing the "lo" byte in the global source map, and to
+    /// make our "span" works in there, we simply add `file_span_lo` to our values.
+    crate fn new_in_file(file_span_lo: u32, lo: u32, hi: u32) -> Self {
+        Self { lo: lo + file_span_lo, hi: hi + file_span_lo }
+    }
+
+    crate fn empty() -> Self {
```

Comment:
Please name this dummy() for consistency.

Reply:
:+1: